### PR TITLE
JSON array fixes

### PR DESCRIPTION
**Pull request by mbolivar**
Fixes and regression tests for the JSON library's array handling.

### CODE REVIEW

**Comment by mbolivar** [comment]
@lpereira got time for another JSON review?

File: lib/json/json.c [hunk 1]
```diff
@@ -713,7 +718,23 @@ static int arr_encode(const struct json_obj_descr *descr, const void *field,
...
+		 * length field in the "parent" struct containing the
+		 * array.
+		 *
+		 * To patch things up, we lie to encode() about where
+		 * the field is by exactly the amount it will offset
+		 * it. This is a size optimization for struct
+		 * json_obj_descr: the alternative is to keep a
+		 * separate field next to element_descr which is an
+		 * offset to the length field in the parent struct,
+		 * but that would add a size_t to every descriptor.
```
Comment (lpereira) [inline lib/json/json.c]: Thanks for adding these comments.

File: lib/json/json.c [hunk 2]
```diff
@@ -518,16 +518,16 @@ static int arr_parse(struct json_obj *obj,
...
 		if (decode_value(obj, elem_descr, &value, field, val) < 0) {
 			return -EINVAL;
 		}
 
 		(*elements)++;
-
 		field = (char *)field + elem_size;
-		if (field == last_elem) {
-			return -ENOSPC;
-		}
```
Comment (lpereira) [inline lib/json/json.c]: Might be nice if we had a test for the case you're fixing with this commit.

File: lib/json/json.h [hunk 3]
```diff
@@ -173,6 +173,61 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
...
+ *          size_t heights_len;
+ *      };
+ *
+ *      struct json_obj_descr person_height_descr[] = {
+ *           JSON_OBJ_DESCR_PRIM(struct person_height, name, JSON_TOK_STRING),
+ *           JSON_OBJ_DESCR_PRIM(struct person_height, height, JSON_TOK_NUMBER),
+ *      };
+ *
+ *      struct json_obj_descr array[] = {
+ *           JSON_OBJ_DESCR_OBJ_ARRAY(struct people_heights, heights, 10,
```
Comment (lpereira) [inline lib/json/json.h]: Where is the field name in this example code?

File: lib/json/json.h [hunk 4]
```diff
@@ -173,6 +173,61 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
...
+ * @param max_len_ Maximum number of elements in the array
+ *
+ * @param len_field_ Field name in the struct for the number of elements
+ * in the array
+ *
+ * @param elem_descr_ Element descriptor, pointer to a descriptor array
+ *
+ * @param elem_descr_len_ Number of elements in elem_descr_
+ *
+ * Here's an example of use:
```
Comment (lpereira) [inline lib/json/json.h]: Can we mark the example of use with some markup that'll get it syntax-highlighted in the documentation?

**Review by lpereira — CHANGES_REQUESTED** [review]
The patches look good, thanks! Just a few minor nits.

File: lib/json/json.c (hunk 2 above)
Comment (mbolivar) [inline lib/json/json.c]: We do -- the last patch adds a test which fills an array completely.

File: lib/json/json.h (hunk 4 above)
Comment (mbolivar) [inline lib/json/json.h]: None of the existing macros' documentation has such markup -- can that be done as a separate cleanup patch, outside of this series?

Comment (lpereira) [inline lib/json/json.h]: Of course.

File: lib/json/json.h (hunk 3 above)
Comment (mbolivar) [inline lib/json/json.h]: The field name is "heights", which is argument 2.

The docstring for JSON_OBJ_DESCR_OBJ_ARRAY() below shows a field name "people-heights" and a struct field name "heights".

File: lib/json/json.h (hunk 4 above)
Comment (mbolivar) [inline lib/json/json.h]: Thanks!

**Comment by mbolivar** [comment]
@lpereira, any other feedback?

**Review by lpereira — APPROVED** [review]
LGTM

**Pull request merged**